File: Top_100_Liked_Questions/139._Word_Break/solution.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        hashSet = set(wordDict)
        mem = [False] * (len(s)+1)
        mem[0] = True
        
        for i in range(1,len(mem)):
            for j in range(i):
                if mem[j] and s[j:i] in hashSet:
                    mem[i] = True
                    break
        
        return mem[len(s)]
    
    
    def wordBreak2(self, s: str, wordDict: List[str]) -> bool:
        wd = sorted(wordDict, key=lambda x: len(x), reverse=True)
        def wb(s):
            if len(s) <= 0:
                return True

            for w in wd:
                if s.startswith(w):
                    if wb(s[len(w):]):
                        return True

            return False
        return wb(s)

File: Top_100_Liked_Questions/139._Word_Break/solution.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function(s, wordDict) {
    let hashSet = {};
    for (let word of wordDict) {
        hashSet[word] = null;
    }
    
    let mem = new Array(s.length+1).fill(false);
    mem[0] = true;
    
    for (let i=1; i<mem.length; i++) {
        for (let j=0; j<i; j++) {
            if (mem[j] && hashSet.hasOwnProperty(s.slice(j,i))) {
                mem[i] = true;
                break;
            }
        }
    }
    
    return mem[s.length];
}


var wordBreak2 = function(s, wordDict) {
    if (s.length <= 0) {
        return true;
    }
    
    for (let word of wordDict) {
        if (startsWith(s, word)) {
            if (wordBreak(s.slice(word.length), wordDict)) {
                return true;
            }
        }
    }
    return false;
};

let startsWith = function(s, i, word) {
    for (let j=0; j<word.length; j++) {
        if (s[i+j] != word[j]) {
            return false;
        }
    }
    return true;
}